### PR TITLE
Fix cmp_cli_test issue 13494

### LINKED ISSUE #13494: Issues in cmp_cli_test
The cmp_cli_test seems to fail in the minimal Github CI build:

https://github.com/openssl/openssl/pull/13489/checks?check_run_id=1443890824

There seems to be at least 2 problems. This code in the cmp app looks very suspect:

https://github.com/openssl/openssl/blob/9524a3089c0d4c3ae553e86f877eef9d6b192b20/apps/cmp.c#L2699-L2707

To start with `ui_fallback_method` does not seem to be used for anything beyond this block. Secondly the `UI_method_set_reader()` function call seems to be modifying the built-in OpenSSL method directly. This works for `UI_OpenSSL()` which is not declared `const` (perhaps it should be?), but not for `UI_null()` which *is* declared const. This latter problem results in a seg-fault.

A second issue is that in the minimal build we have `no-des` specified. Unfortunately there is a key used in the test that seems to be encrypted with DES, and hence this fails.

### PR DESCRIPTION
**Pull request by DDvO**
* `apps/cmp.c`: fix crash with `-batch` option on `OPENSSL_NO_UI_CONSOLE`
* re-encrypt `81-test_cmp_cli_data/Mock/signer.p12` using AES-256-CBC (avoiding DES)

Fixes #13494

### CODE REVIEW

File: apps/cmp.c [hunk 1]
```diff
@@ -2697,13 +2697,13 @@ int cmp_main(int argc, char **argv)
...
-        UI_METHOD *ui_fallback_method;
 #ifndef OPENSSL_NO_UI_CONSOLE
-        ui_fallback_method = UI_OpenSSL();
-#else
-        ui_fallback_method = (UI_METHOD *)UI_null();
+        /*        static UI_METHOD ui_null = {"NULL UI",
+                                    NULL, NULL, NULL, NULL, NULL, NULL};
+        UI_METHOD *ui_fallback_method = (UI_METHOD *)UI_null();
+        */
+        UI_method_set_reader(UI_OpenSSL(), NULL);
```
Comment (mattcaswell) [inline apps/cmp.c:2705]: Should we really be "setting" this OpenSSL global method? Shouldn't it be declared as const to prevent this?
I was expecting a solution that copied the UI_METHOD instead. In fact this code looks very much like this existing code where we do make a copy:

https://github.com/openssl/openssl/blob/9524a3089c0d4c3ae553e86f877eef9d6b192b20/apps/lib/apps_ui.c#L115-L129

Comment (mattcaswell) [reply apps/cmp.c:2705]: > Shouldn't it be declared as const to prevent this?

Although changing the return value of `UI_OpenSSL()` itself is probably an API break...so maybe not.

**Comment by kroeckx** [comment]
I think this fixes the minimal CI test that's failing for other pull requests, so I put an urgent label on it.

**Comment by DDvO** [comment]
The sanitizers CI failure is unrelated.

File: apps/cmp.c (hunk 1 above)
Comment (DDvO) [reply apps/cmp.c]: The code adapting the UI method `cmp.c` when the `-batch` option is given had indeed been borrowed from an earlier version of the above quoted code in `apps_ui.c`.

The goal is to disable any interactive app input, and this so far has been achieved by changing the UI method globally. To admit, this was a pretty invasive way of doing that. 
I've just had a closer look and found a more elegant way, namely to modify only the app-specific UI method. This required removing `const` from the declaration of `UI_METHOD *get_ui_method(void)`, which I believe is adequate.

**Comment by kroeckx** [comment]
> The sanitizers CI failure is unrelated.

I've created #13503 for that.

**Comment by DDvO** [comment]
Even in default builds for some reason there was a further issue reading encrypted PKCS#12 files:
```
Could not read private key for CMP client certificate from signer.p12
C007A9C9437F0000:error::STORE routines:(unknown function):error verifying pkcs12 mac:crypto/store/store_result.c:563:
```
I tried finding out why, but did not succeed so far.
In the interest of quickly fixing the CI runs with `cmp_clit_test` I've switched to using PEM input (`signer.key`) for now.

**Comment by mattcaswell** [comment]
Hmm CI failures still...

File: apps/cmp.c (hunk 1 above)
Comment (mattcaswell) [reply apps/cmp.c]: I'm still slightly confused....where is the ui method actually *used* by the cmp app?

Comment (DDvO) [reply apps/cmp.c]: Sigh, it turns out that the more elegant version does not work because `load_key_certs_crls()` relies on the UI method returned by `get_ui_method()`, so changed the fix back to adapting `UI_OpenSSL()` instead.

**Comment by DDvO** [comment]
> Hmm CI failures still...

This should be fixed now; see my above comment on your review.

BTW, rebasing on the current master lead to an unrelated build failure on `test/pem_read_depr_test`
So I've provisionally rebased this PR on the slightly earlier commit ee8252847d1ccf39a48ab382971f8f18ac3f4089.

File: apps/cmp.c (hunk 1 above)
Comment (DDvO) [reply apps/cmp.c:2702]: > where is the ui method actually used by the cmp app?

The UI methods are used by all apps using `load_key()` and the like, which rely on `load_key_certs_crls()`.

Comment (t8m) [reply apps/cmp.c:2702]: I suppose the clean fix would be to to add set_batch_ui_method() which would set the NULL reader. in the ui_method returned by get_ui_method?

File: apps/lib/apps_ui.c [hunk 2]
```diff
@@ -136,7 +136,7 @@ void destroy_ui_method(void)
     }
 }
 
-const UI_METHOD *get_ui_method(void)
+UI_METHOD *get_ui_method(void)
```
Comment (mattcaswell) [inline apps/lib/apps_ui.c:139]: Should we revert this change too?

Comment (DDvO) [reply apps/lib/apps_ui.c:139]: I suggest leaving this in because it may be useful in other situations.

File: apps/cmp.c (hunk 1 above)
Comment (DDvO) [reply apps/cmp.c:2702]: As I tried to explain above removing the reader method from the apps' `ui_method` (regardless which way this is done) is not the right solution because `load_key_certs_crls()` unfortunately uses this UI method for passing further down the `pass` argument. So any given password is lost this way :-(

Comment (t8m) [reply apps/cmp.c:2702]: Ah, now I get it. But you could then set a flag to just return null and not call the reader from ui_fallback_method in case the password is not set.

Comment (DDvO) [reply apps/cmp.c:2702]: The mentioned problem is in case a password must be used, not when it is not set).

I suggest doing any further discussion on this later 
because getting this fix merged is urgent to fix CI test failures
and I have other very urgent things to do today.

**Review by mattcaswell — APPROVED** [review]
I'm not too enthusiastic about the UI_METHOD stuff. But my concern there pre-dates this PR, and this PR improves the CI situation. So approved.

**Comment by DDvO** [comment]
> I'm not too enthusiastic about the UI_METHOD stuff. But my concern there pre-dates this PR, and this PR improves the CI situation. So approved.

Merged already now since this was urgent.

Shall I keep this PR open for now as a reminder to finish the above discussion?

**Comment by mattcaswell** [comment]
> Shall I keep this PR open for now as a reminder to finish the above discussion?

I raised issue #13511 for this, so I think this PR can be closed.
